### PR TITLE
[KEYCLOAK-19311] DistX integration tests on Testcontainers

### PR DESCRIPTION
**Pull request by andreaTP**
This PR improves the Dist.X Integration tests and add a switch to run them on Testcontainers too (with Keycloak spinned up as a Docker container).

### CODE REVIEW

File: quarkus/tests/integration/src/main/java/org/keycloak/it/utils/DockerKeycloakDistribution.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+    private File distributionFile = new File("../../../distribution/server-x-dist/target/keycloak.x-" + Version.VERSION_KEYCLOAK + ".tar.gz");
+
+    private GenericContainer keycloakContainer = null;
+
+    private GenericContainer runKeycloakContainer() {
+        if (!distributionFile.exists()) {
+            throw new RuntimeException("Distribution archive " + distributionFile.getAbsolutePath() +" doesn't exists");
+        }
+        // TODO: decide on one approach
```
Comment (andreaTP) [inline quarkus/tests/integration/src/main/java/org/keycloak/it/utils/DockerKeycloakDistribution.java:40]: Here I'm providing the two implementations:
 - using a local `Dockerfile`
 - downloading the server-x `Dockerfile` from GH
 
@stianst can we consider to move the server-x `Dockerfile` to only this repository in the future?

I'm happy to keep only the preferred and agreed implementation BTW.

**Review by pedroigor — COMMENTED** [review]
@andreaTP Now I remember another reason for resolving the distribution artifact via Maven ...

It avoids having to build the dist module but relies on previous artifacts installed into the local repo. That seems to be related to the GHA failure.

I'm OK if you just change the `Quarkus Tests` step to build the distribution prior to running the tests.

Otherwise, LGTM!

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -403,6 +403,14 @@ jobs:
           java-version: ${{ env.DEFAULT_JDK_VERSION }}
       - name: Update maven settings
         run: mkdir -p ~/.m2 ; cp .github/settings.xml ~/.m2/
+      - name: Prepare the local distribution archives
+        run: mvn clean install -DskipTests -Pdistribution
+      - name: Run Quarkus Tests in Docker
+        run: |
+          mvn clean install -nsu -B -f quarkus/tests/pom.xml -DkcDistribution=docker | misc/log/trimmer.sh
```
Comment (pedroigor) [inline .github/workflows/ci.yml:410]: Shall we keep the format we are using for properties and have `kcDistribution` as `kc.tests.dist` ? Or something similar.

Also, I'm wondering if it makes more sense to run tests in GHA against docker, only. It should be enough to test not only the image but the distribution itself (as we do for the standalone distribution).

File: quarkus/tests/integration/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,34 @@
+FROM registry.access.redhat.com/ubi8-minimal AS build-env
```
Comment (pedroigor) [inline quarkus/tests/integration/Dockerfile:1]: Do we still need this file? Aren't you now fetching the docker file via HTTP?

File: .github/workflows/ci.yml (hunk 2 above)
Comment (andreaTP) [reply .github/workflows/ci.yml:410]: I'm happy with whatever name you prefer 🙂

File: quarkus/tests/integration/Dockerfile (hunk 3 above)
Comment (andreaTP) [reply quarkus/tests/integration/Dockerfile:1]: As mentioned, I kept the two implementations until a final decision is made.

**Comment by andreaTP** [comment]
> I'm OK if you just change the Quarkus Tests step to build the distribution prior to running the tests.

The Quarks tests actually passed in Ci using both Docker and Raw implementations.

Docker tests are (at the moment) much slower because we are rebuilding images and waiting for timeouts.
For the moment I would run both the distributions in CI.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (DGuhr) [reply .github/workflows/ci.yml:410]: Lets stay consistent then, i'd say :)

**Comment by DGuhr** [comment]
Btw, what about the test setup w.r.t. unit tests? Seems these tests or some of the quarkus integration tests also get executed there, which causes the step to fail.

**Pull request closed without merging**